src/org/traccar/protocol/GranitProtocolDecoder.java

```diff
@@ -0,0 +1,215 @@
+        response.writeBytes("BB+ARCF~".getBytes(StandardCharsets.US_ASCII));
+        response.writeShort(4); //binary length
+        response.writeShort(packNum);
+        response.writeShort(deviceId);
+        appendChecksum(response, 14);
+        channel.write(response);
+    }
+
+    private static void decodeStructure(ChannelBuffer buf, Position position) {
+        short flags = buf.readUnsignedByte();
+        position.setValid(BitUtil.check(flags, 7));
+        position.set(Position.KEY_ALARM, BitUtil.check(flags, 1));
+
+        short satDel = buf.readUnsignedByte();
+        position.set(Position.KEY_SATELLITES, BitUtil.from(satDel, 4));
+
+        int pdop = BitUtil.to(satDel, 4);
+        position.set("pdop", pdop);
+
+        int lonDegrees = buf.readUnsignedByte();
+        int latDegrees = buf.readUnsignedByte();
+        int lonMinutes = buf.readUnsignedShort();
+        int latMinutes = buf.readUnsignedShort();
+        double latitude = latDegrees + (double) latMinutes / 60000;
+        double longitude = lonDegrees + (double) lonMinutes / 60000;
+        if (!BitUtil.check(flags, 4)) {
+            latitude = -latDegrees;
+        }
+        if (!BitUtil.check(flags, 5)) {
+            longitude = -longitude;
+        }
+        position.setLongitude(longitude);
+        position.setLatitude(latitude);
+
+        position.setSpeed(buf.readUnsignedByte());
+
+        int course = buf.readUnsignedByte();
+        if (BitUtil.check(flags, 6)) {
+            course = course | 0x100;
+        }
+        position.setCourse(course);
+
+        position.set(Position.KEY_DISTANCE, buf.readShort());
+
+        int analogIn1 = buf.readUnsignedByte();
+        int analogIn2 = buf.readUnsignedByte();
+        int analogIn3 = buf.readUnsignedByte();
+        int analogIn4 = buf.readUnsignedByte();
+
+        int analogInHi = buf.readUnsignedByte();
+
+        analogIn1 = analogInHi << 8 & 0x300 | analogIn1;
+        analogIn2 = analogInHi << 6 & 0x300 | analogIn2;
+        analogIn3 = analogInHi << 4 & 0x300 | analogIn3;
+        analogIn4 = analogInHi << 2 & 0x300 | analogIn4;
+
+        position.set(Position.PREFIX_ADC + 1, analogIn1);
+        position.set(Position.PREFIX_ADC + 2, analogIn2);
+        position.set(Position.PREFIX_ADC + 3, analogIn3);
+        position.set(Position.PREFIX_ADC + 4, analogIn4);
+
+        position.setAltitude(buf.readUnsignedByte() * 10);
+
+        short diOut = buf.readUnsignedByte();
+        for (int i = 0; i < 8; i++) {
+            position.set(Position.PREFIX_IO + (i + 1), BitUtil.check(diOut, i));
+        }
+        buf.skipBytes(1); //StatMess
+    }
+
+    @Override
+    protected Object decode(Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        ChannelBuffer buf = (ChannelBuffer) msg;
+
+        if (hasDeviceId() &&  buf.toString(StandardCharsets.US_ASCII).contains("OK")
```

Comment:
I think it would be more efficient and cleaner to call `buf.toString(StandardCharsets.US_ASCII)` once.